### ISSUE TITLE
check for default config path

### ISSUE THREAD

**Issue by Setsua** [Mydayyy/pbcli#17]
I want to deploy your program on multiple hosts but setting up an environment variable for each user is not very convenient. 
Can you implement default paths with different weightings?

Something like this:
PBCLI_CONFIG_PATH > $HOME/.config/pbcli/config > /etc/pbcli/config

The environment variable has priority over the user config folder which in turn has priority over the system path.

**Comment by Mydayyy** [comment]
Greetings,

Thank you for the feature request. Initially I took inspiration from how ripgrep handles the configuration of their CLI tool and they explicitly do not have a default config location. However, having one is a common practice in many other tools and I see no reason to not implement it if there is a need for it. I'll try to work on this on the weekend and would implement the following:

- Default fallback locations for configs with the environment variable always taking precedence
- A new cli option named `--no-default-config` to disable looking for a default config

Best Regards
Mydayyy

**Comment by Mydayyy** [comment]
Greetings,

I apologize for the delay. I finished up this feature and will proceed to merge and release this tomorrow since its currently fairly late. I now implemented the following three things:

- A new cli optioned as mentioned above named `--no-default-config`. This will prevent the application from looking around for a default config but will still respect the config set via the envvar
- A new logger mechanism. Currently only used to output config related messages (and other messages from libraries used). It can be activated with `--debug` and should be useful to troubleshoot config issues, for example seeing which directories are searched and which config is loaded<sup>1</sup>
- Default config locations. The searched paths depend on the system and follow os-specific conventions. On linux this should be `$XDG_CONFIG_HOME/pbcli/config` or `$HOME/.config/pbcli/config` if the XDG variable is not set. Additionally, `/etc/pbcli/config` will also be searched. For reference: https://docs.rs/directories/latest/directories/struct.ProjectDirs.html#method.config_local_dir



<sup>1</sup> Example output:
```DEBUG pbcli::config: PBCLI_CONFIG_PATH not set
DEBUG pbcli::config: looking for config at /home/mydayyy/.config/pbcli/config
DEBUG pbcli::config: looking for config at /etc/pbcli/config
DEBUG pbcli::config: no config found
```


Edit: Pushed here and on crates.io. Leaving this issue open until I updated the AUR tomorrow.

**Comment by Mydayyy** [comment]
Released in 2.5.0 on crates, aur and github

**Comment by Ilav1** [comment]
Hi @Mydayyy ,

I thing I need some help with the configuration file. :-(

pbcli is stored in `/etc/pbcli/`. 
`echo 'TestPaste' | /etc/pbcli/pbcli --host https://privatebin.net/ --burn` is working fine. Now I want to use the config file.
I created the config in `/etc/pbcli/ ('/etc/pbcli/config')` and add the following lines:

```
--host=https://privatebin.net/
--burn
```

Now the comand should work withour `--host https://privatebin.net/ --burn`

`echo 'TestPaste' | /etc/pbcli/pbcli` does not work.



```
#echo 'test' | /etc/pbcli/pbcli
              error: The following required arguments were not provided:
    <URL>

USAGE:
    pbcli <URL>

For more information try --help
```


`echo 'TestPaste' | /etc/pbcli/pbcli --debug` does not work.

```
#echo 'test' | /etc/pbcli/pbcli --debug
              error: The following required arguments were not provided:
    <URL>

USAGE:
    pbcli <URL>

For more information try --help
```

`echo 'TestPaste' | /etc/pbcli/pbcli --host https://privatebin.net/ --burn --debug` is working fine.

The debug report is using the config `/etc/pbcli/config`

```
# echo 'test' | /etc/pbcli/pbcli --host https://privatebin.net/ --burn --debug
DEBUG pbcli::config: PBCLI_CONFIG_PATH not set
DEBUG pbcli::config: looking for config at /root/.config/pbcli/config
DEBUG pbcli::config: looking for config at /etc/pbcli/config
DEBUG pbcli::config: using config /etc/pbcli/config
DEBUG reqwest::connect: starting new connection: https://privatebin.net/
DEBUG hyper_util::client::legacy::connect::dns: resolving host="privatebin.net"
DEBUG hyper_util::client::legacy::connect::http: connecting to [2001:1620:2057:1:4902::7]:443
DEBUG hyper_util::client::legacy::connect::http: connected to [2001:1620:2057:1:4902::7]:443
DEBUG hyper_util::client::legacy::pool: pooling idle connection for ("https", privatebin.net)
https://privatebin.net/?f50d5e0825430de6#8TzwbEyCa3AztCBErt37tYmi8YJ3r48uTUvmVuFRRbSi
```


Now I set `PBCLI_CONFIG_PATH=/etc/pbcli/config`.


`echo 'TestPaste' | /etc/pbcli/pbcli --debug` does not work.

```
#echo 'test' | /etc/pbcli/pbcli --debug
              error: The following required arguments were not provided:
    <URL>

USAGE:
    pbcli <URL>

For more information try --help
```

`echo 'TestPaste' | /etc/pbcli/pbcli --host https://privatebin.net/ --burn --debug` is working fine.

The debug report is using the config `/etc/pbcli/config` again


```
# echo 'test' | /etc/pbcli/pbcli --host https://privatebin.net/ --burn --debug --no-default-config
DEBUG pbcli::config: using config pointed to by PBCLI_CONFIG_PATH
DEBUG pbcli::config: using config /etc/pbcli/config
DEBUG reqwest::connect: starting new connection: https://privatebin.net/
DEBUG hyper_util::client::legacy::connect::dns: resolving host="privatebin.net"
DEBUG hyper_util::client::legacy::connect::http: connecting to [2001:1620:2057:1:4902::7]:443
DEBUG hyper_util::client::legacy::connect::http: connected to [2001:1620:2057:1:4902::7]:443
DEBUG hyper_util::client::legacy::pool: pooling idle connection for ("https", privatebin.net)
https://privatebin.net/?068f39ac53387327#AXaJt1RWQHWJ5oEjLabrjHRaZGxvVSEczzhpG9FQoByW
```




I dont know how to use the config the right way. Im using ubunti 24.04.

Maybee somone can help me?

**Comment by Mydayyy** [comment]
Greetings @Ilav1,

thank you for your report. I would kindly ask you to open a new issue (you can just copy paste this comment, it is very thorough) so we do not spam watchers of this closed issue while we figure this out. In the meantime I will look into this and re-test it on my end.


Best Regards
Mydayyy

**Comment by Mydayyy** [comment]
Greetings,

the fix has been released as v2.5.1 on github, AUR and on crates. Please let me know if this works for you and I will proceed to close this issue.

Best Regards
Mydayyy
